etc/scripts/analysis/user-activity.js

```diff
@@ -0,0 +1,510 @@
+
+            // Re-group things into sane concurrent batches
+            .groupBy(function(userHash, i) { return Math.floor(i / 5); })
+            .values()
+            .value();
+
+        log().info('Begin writing %s batches of user activity to CSV', batches.length);
+
+        return _writeUserRows(batches);
+    });
+});
+
+/**
+ * Build and write all user rows concurrently for the given batches of user hashes
+ *
+ * @param  {Object[][]}     userHashBatches     A collection of batches for which to generate user rows and write them to the csv file
+ * @api private
+ */
+function _writeUserRows(userHashBatches) {
+    if (_.isEmpty(userHashBatches)) {
+        log().info('Processing complete');
+        return _exit(0);
+    }
+
+    log().info('There are %d batches remaining to process', userHashBatches.length);
+
+    // Process the next batch
+    var userHashBatch = userHashBatches.shift();
+    var _done = _.after(userHashBatch.length, function() {
+        log().info('Successfully completed batch of %d users', userHashBatch.length);
+        return _writeUserRows(userHashBatches);
+    });
+
+    _.each(userHashBatch, function(userHash) {
+        _createUserRow(userHash, function(err, row) {
+            if (err) {
+                log().warn({'err': err, 'user': userHash}, 'Failed to create row from a user hash');
+                return _done();
+            }
+
+            return csvStream.write(row, _done);
+        });
+    });
+}
+
+/**
+ * Page through all the users in the system, filtering by tenant and those that have email
+ * addresses, and group them by their email address
+ *
+ * @param  {String}     tenantAlias             The tenant alias to filter by
+ * @param  {Function}   callback                Invoked when users have been grouped by email address
+ * @param  {Object}     usersByEmail            An object keyed by email, whose value is an array of user hashes
+ * @api private
+ */
+function _groupUsersByEmail(tenantAlias, callback) {
+    var userHashes = [];
+
+    PrincipalsDAO.iterateAll(['principalId', 'tenantAlias', 'displayName', 'email', 'visibility', 'lastModified', 'notificationsLastRead'], 100, _aggregateUsers, function(err) {
+        if (err) {
+            log().error({'err': err}, 'Failed to iterate all users');
+            return process.exit(1);
+        }
+
+        log().info('Finished indexing %s users by email', userHashes.length);
+
+        return callback(_.groupBy(userHashes, 'email'));
+    });
+
+    /*!
+     * Given a paged result set of principal hashes, filter them down to those with email addresses
+     * and those that are part of the specified tenant. Then add them to the shared `userHashes`
+     * array
+     *
+     * @param  {Object[]}   principalHashes     The principals to filter and aggregate
+     * @param  {Function}   callback            Will be invoked when the principals are aggregated
+     */
+    function _aggregateUsers(principalHashes, callback) {
```

Comment:
> Is there a reason to deviate from the style guide other than stylistic preference?

The style guide is built on the assumption of using module-based invocation, where there is a run of the code first before any actual method invocation is performed (e.g., when you do your first `require`, all you're normally doing is declaring your variables / functions in the file and exporting functionality, not usually invoking actual functionality). Because of this, we can define public functions at the top, and private functions at the bottom, which I believe is part of our stylistic guideline (and a more significant guideline than exactly how functions are declared).

In these scripts, the environment is different as the first and only invocation is the one that is invoking functions. We have to either make a decision to violate our style guidelines by declaring all the very trivial internal functionality at the top (e.g., a function to split a string in a certain way), or violate it by just declaring the functions using `function`, and I chose to declare the functions differently because every time I open up a script I don't want to have to scroll down to the bottom to find the substance of the script functionality.

So yes it was a choice based on stylistic preference, but we were going to have to violate "style guidelines" some how.